mlir/include/mlir/Tools/mlir-opt/MlirOptMain.h

```diff
@@ -176,6 +189,10 @@ class MlirOptMainConfig {
   /// general.
   bool allowUnregisteredDialectsFlag = false;
 
+  /// The callback for additional configuration of the MLIR context after it has
+  /// been created and populated from the registry.
+  std::function<LogicalResult(MLIRContext &)> contextConfigurationCallback;
+
```

Comment:
This isn't the first time his came up, and we've always heavily pushed back on adding unconstrained MLIRContext access here.

I need to review this more carefully.

Reply:
Fwiw, I've been owning these tools downstream for years and have been on the verge of forking away from optmain many times because it is too brittle. I think we were wrong on being so restrictive on it, and that is a pretty strong opinion based on experience at this point. 

But I'm not going to spend the energy to argue it very much in the abstract: it's just a utility/sample code and I'll probably fork it some day if needs don't match upstream expectations.

Reply:
It's perfectly fine to fork a restrictive API if it you have exotic use-cases that don't fit the abstraction.

The OptMain API has evolved based on the service intended to be provided by `*-opt` tools, and the experience from setting up many of them. Ultimately we consolidate behind some formal API with a config abstraction that was enough to support all the known cases (while providing a "safe" abstraction to set up consistent tool).

This restrictive API is wrapping lower level APIs, which all remain available to any tool that isn't fitting the `mlir-opt` contract (back to forking?)